File: server/modules/git/status.ts
export type GitFileStatusCode =
  | "added"
  | "modified"
  | "deleted"
  | "renamed"
  | "copied"
  | "untracked"
  | "unmerged"
  | "ignored"
  | "unknown";

export interface GitStatusEntry {
  path: string;
  previousPath?: string;
  staged: GitFileStatusCode | null;
  unstaged: GitFileStatusCode | null;
}

function mapStatusChar(char: string): GitFileStatusCode | null {
  switch (char) {
    case " ":
      return null;
    case "?":
      return "untracked";
    case "!":
      return "ignored";
    case "A":
      return "added";
    case "M":
      return "modified";
    case "D":
      return "deleted";
    case "R":
      return "renamed";
    case "C":
      return "copied";
    case "U":
      return "unmerged";
    default:
      return "unknown";
  }
}

/** Parse `git status --porcelain` output into structured entries. */
export function parseGitStatusPorcelain(output: string): GitStatusEntry[] {
  const entries: GitStatusEntry[] = [];
  for (const line of output.split("\n")) {
    if (!line.trim()) continue;
    const indexStatus = line[0] ?? " ";
    const worktreeStatus = line[1] ?? " ";
    let pathPart = line.slice(3);

    let previousPath: string | undefined;
    const renameArrow = " -> ";
    if (pathPart.includes(renameArrow)) {
      const [from, to] = pathPart.split(renameArrow);
      previousPath = from?.trim();
      pathPart = to?.trim() ?? pathPart;
    }

    const path = pathPart.trim();
    if (!path) continue;

    const staged = mapStatusChar(indexStatus);
    const unstaged = mapStatusChar(worktreeStatus);
    const status = staged ?? unstaged;
    if (!status) continue;

    entries.push({
      path,
      previousPath,
      staged,
      unstaged,
    });
  }
  return entries;
}

File: server-rs/src/git/status.rs
use crate::types::GitStatusEntry;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GitFileStatusCode {
    Added,
    Modified,
    Deleted,
    Renamed,
    Copied,
    Untracked,
    Unmerged,
    Ignored,
    Unknown,
}

impl GitFileStatusCode {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Added => "added",
            Self::Modified => "modified",
            Self::Deleted => "deleted",
            Self::Renamed => "renamed",
            Self::Copied => "copied",
            Self::Untracked => "untracked",
            Self::Unmerged => "unmerged",
            Self::Ignored => "ignored",
            Self::Unknown => "unknown",
        }
    }
}

fn map_status_char(ch: char) -> Option<GitFileStatusCode> {
    match ch {
        ' ' => None,
        '?' => Some(GitFileStatusCode::Untracked),
        '!' => Some(GitFileStatusCode::Ignored),
        'A' => Some(GitFileStatusCode::Added),
        'M' => Some(GitFileStatusCode::Modified),
        'D' => Some(GitFileStatusCode::Deleted),
        'R' => Some(GitFileStatusCode::Renamed),
        'C' => Some(GitFileStatusCode::Copied),
        'U' => Some(GitFileStatusCode::Unmerged),
        _ => Some(GitFileStatusCode::Unknown),
    }
}

fn code_to_option_str(code: Option<GitFileStatusCode>) -> Option<String> {
    code.map(|c| c.as_str().to_string())
}

/// Parse `git status --porcelain` output into structured entries.
pub fn parse_git_status_porcelain(output: &str) -> Vec<GitStatusEntry> {
    let mut entries = Vec::new();

    for line in output.lines() {
        if line.trim().is_empty() {
            continue;
        }

        let chars: Vec<char> = line.chars().collect();
        let index_status = chars.first().copied().unwrap_or(' ');
        let worktree_status = chars.get(1).copied().unwrap_or(' ');
        let mut path_part = if line.len() >= 3 {
            line[3..].to_string()
        } else {
            String::new()
        };

        let mut previous_path: Option<String> = None;
        const RENAME_ARROW: &str = " -> ";
        if path_part.contains(RENAME_ARROW) {
            if let Some((from, to)) = path_part.split_once(RENAME_ARROW) {
                previous_path = Some(from.trim().to_string());
                path_part = to.trim().to_string();
            }
        }

        let path = path_part.trim().to_string();
        if path.is_empty() {
            continue;
        }

        let staged = map_status_char(index_status);
        let unstaged = map_status_char(worktree_status);
        let status = staged.or(unstaged);
        if status.is_none() {
            continue;
        }

        entries.push(GitStatusEntry {
            path,
            previous_path,
            staged: code_to_option_str(staged),
            unstaged: code_to_option_str(unstaged),
        });
    }

    entries
}
